### ISSUE TITLE
Loners refuse to take Bandit's Base

### ISSUE THREAD

**Issue by AlexDemolisher** [Decane/SRP#23]
Loners refuse to take bandit's base despite having "heading to depot", instead occupying spots near it

**Comment by SurDno** [comment]
Are you sure this issue persists with the current version of SRP? I know it was a problem in vanilla game, but SRP Faction War works flawlessly for me.
Too little details.

**Comment by AlexDemolisher** [comment]
SRP 1.1.2. I had opposite experience, vanilla 1.5.10 loners always took bandit base without problem, but in SRP they only take places nearby. I managed to take bandit's base by going with loners to nearby post and then initiating fight between depot bandits and loner squad.

**Comment by SurDno** [comment]
You haven't met it in vanilla? You're lucky. Just go search Steam Discussions, GSC Forums, Mail.ru Answers, anything. There are thousands of people with this bug, who are waiting for Loners' squad to arrive for days. And, yep, it still persists with 1.5.10 patch.

Since this issue is already stated as fixed, I doubt Decane can magically understand what's your problem. Upload your savegame on some kind of file-sharing site and then give a link here.

**Comment by Decane** [comment]
The faction war is not entirely fixed in SRP v1.1.2. AFAIK, it _is_ in v1.1.3, which fixes the last remaining problem I knew to sometimes cause the war to grind to a halt in v1.1.2. From the 1.1.3 readme:

>Implemented a tail-recursive algorithm to automatically detect and resolve faction war stalemates arising out of a squad being grid-locked into a camp by all surrounding camps being occupied by non-hostiles, thereby preventing the grid-locked squad from leaving the camp.

So I don't believe there is anything to do here.

However, I can have a look at your savegame, @AlexDemolisher, to see if the bug afflicting it really is the one fixed by the algorithm above, if you upload it as SurDno suggested.

**Comment by AlexDemolisher** [comment]
Here's one of the saves: https://mega.nz/#!lwtFHQZB!k4sbqEy14KeP3yEArcKDf1xdhbuPB3OqeZM_-lcDM-k
Does 1.1.3 require new game if I'm on 1.1.2?
Anyway, thanks for your work.

**Comment by Decane** [comment]
Thank you for uploading your savegame, @AlexDemolisher. I inspected it and, in doing so, discovered yet another vanilla bug that can block the faction war. I now recognize that @Marek33's savegame was also afflicted with this bug a while back.

The reason your faction war is stuck is because there are several rogue 'capture' tasks in your `task_manager.get_task_manager().tasks_by_smarts` table:
```
! [LUA][ERROR] tm.tasks_by_smarts: smart = Boat park, name = sim_capture_smart, entity_id 52, status normal
! [LUA][ERROR] tm.tasks_by_smarts: smart = Hideout behind pipes, name = sim_capture_smart, entity_id 171, status normal
! [LUA][ERROR] tm.tasks_by_smarts: smart = Encampment near truck, name = sim_capture_smart, entity_id 64, status normal
! [LUA][ERROR] tm.tasks_by_smarts: smart = Gravel pit, name = sim_capture_smart, entity_id 45, status normal
```
One of these has "Hideout behind pipes" as its target and the loner squad closest to the Depot in your game wants to reach the Depot via the "Hideout behind pipes". But because the game mistakenly thinks that's the target of a 'capture' task, it won't let the squad do that.

The only way I can see for those task entries to get into `tasks_by_smarts` during gameplay is for the player to be tasked with capturing the specified camp, but I don't see an entry in your PDA for your having been tasked with capturing the "Hideout behind pipes" (Укрытие за трубами), which the table references. So the task can't have made it into the table during gameplay. That means it must have appeared in there during a save/load operation.

Interestingly, the rogue tasks also have entries in the `inited_tasks` table. And in task_manager.script's `CRandomTask:load()` function, I see that any tasks saved into `inited_tasks` are inserted by the game into `tasks_by_smarts` during a load, regardless of whether the said tasks were in progress (i.e. with status "selected") or not. And a 'capture' task will go into `inited_tasks` during normal gameplay whenever the player asks a squad leader for work and that squad leader offers a 'capture' task.

So, here is what I believe happened: you asked a squad leader for work and he offered a job to capture the "Hideout behind pipes", but you did not accept it. The task went into `inited_tasks` (as it should) and you saved your game. You then loaded this save and the game copied the task from `inited_tasks` to `tasks_by_smarts`, which threw off your faction war. I doubt the copy operation should have happened, but to decide that, I need to check whether every use of `tasks_by_smarts` by the game assumes that it contains only active tasks.

Any which way, this is a genuine bug.

**Comment by Decane** [comment]
Fixed in commit 58699c4.

@AlexDemolisher, to fix your game, find the following lines in your gamedata/scripts/task_manager.script file:
```
if obj.status == "selected" then
	self.task_info[tostring(obj.entity_id)] = obj
end
if (t == "eliminate_smart") or (t == "capture_smart") or (t == "defend_smart") or (t == "defend_smart_delay") then
	TT[obj.target] = obj
end
```
Replace with:
```
if obj.status == "selected" then
	self.task_info[tostring(obj.entity_id)] = obj
	if (t == "eliminate_smart") or (t == "capture_smart") or (t == "defend_smart") or (t == "defend_smart_delay") then
		TT[obj.target] = obj
	end
end
```
That's it.

**Comment by AlexDemolisher** [comment]
Worked 3 times from 3 different saves. Thanks for the fix and your mod.
